host/sampi/state.go

```diff
@@ -117,45 +149,69 @@ outer:
 	}
 	h.Jobs = jobs
 	s.next[hostID] = h
+	l.Debug("marking state as modified")
 	s.nextModified = true
 }
 
 func (s *State) HostExists(id string) bool {
 	_, exists := s.next[id]
+	s.logger.Debug("checking if host exists", "fn", "HostExists", "host.id", id, "exists", exists)
 	return exists
 }
 
 func (s *State) AddHost(host *host.Host, ch chan<- *host.Job) {
+	l := s.logger.New("fn", "AddHost", "host.id", host.ID)
+	l.Debug("adding host")
 	s.next[host.ID] = *host
 	s.streams[host.ID] = ch
+	l.Debug("marking state as modified")
 	s.nextModified = true
 }
 
 func (s *State) RemoveHost(id string) {
+	l := s.logger.New("fn", "RemoveHost", "host.id", id)
+	l.Debug("removing host", "at", "delete.next")
 	delete(s.next, id)
 	s.deleted[id] = struct{}{}
 	if stream := s.streams[id]; stream != nil {
+		l.Debug("closing stream", "at", "close_stream")
 		close(stream)
 	}
+	l.Debug("removing streams", "at", "delete.streams")
 	delete(s.streams, id)
+	l.Debug("marking state as modified")
 	s.nextModified = true
 }
 
 func (s *State) AddListener(ch chan host.HostEvent) {
+	l := s.logger.New("fn", "AddListener")
+	l.Debug("locking listeners")
 	s.listenMtx.Lock()
+	l.Debug("adding listener")
 	s.listeners[ch] = struct{}{}
+	l.Debug("unlocking listeners")
 	s.listenMtx.Unlock()
 }
 
 func (s *State) RemoveListener(ch chan host.HostEvent) {
+	l := s.logger.New("fn", "RemoveListener")
+	l.Debug("locking listeners")
 	s.listenMtx.Lock()
+	l.Debug("removing listener")
 	delete(s.listeners, ch)
+	l.Debug("unlocking listeners")
 	s.listenMtx.Unlock()
 }
 
 func (s *State) sendEvent(hostID, event string) {
+	l := s.logger.New("fn", "sendEvent", "host.id", hostID, "event", event)
+	l.Debug("locking listeners")
 	s.listenMtx.RLock()
-	defer s.listenMtx.RUnlock()
+	defer func() {
+		l.Debug("unlocking listeners")
+		s.listenMtx.RUnlock()
```

Comment:
We don't need to wrap the mutex lock/unlocks in logging calls, that will just slow things down a ton.